app/models/reservation.rb

```diff
@@ -17,4 +17,24 @@ class Reservation < ApplicationRecord # :nodoc:
   validates :checkin_at, presence: true
   validates :checkout_at, presence: true
   validates :book_code, presence: true, uniqueness: true
+  validate :room_must_be_available, if: -> { room_id.present? || checkin_at.present? || checkout_at.present? }
+
+  # Scopes
+  scope :avoid_overlapping_reservations, lambda { |checkin_at, checkout_at, room_id, reservation_id = nil|
+    where('checkin_at < ? AND checkout_at > ?', checkout_at, checkin_at)
+      .where(room_id:)
+      .where.not(id: reservation_id)
+  }
+
+  private
+
+  def room_must_be_available
+    errors.add(:base, 'Room is not available for the selected dates') if Reservation.avoid_overlapping_reservations(checkin_at, checkout_at, room_id, id).exists?
+
+    room = Room.find_by(status: 'available', category: self.room.category)
+
+    errors.add(:status, message: "There's no room with available status, select another category") if room.nil?
+
+    self.room = room
+  end
```

Comment:
I want to clarify what you are trying to do with this method. It sounds to me like that you want to check if the selected room has overlapping reservations, and if it does, then try to find another room in the same category, is that correct?

If that is the case, you can't add error in the first check, the validation then will always fail. I also don't think it should be done as validation. I think this should be a `before_create` method.

You need to decouple the validation from finding an alternative room.